Add support for Direct3D 12 OpenXR backend

mbucchia opened the pull request:
Add support for Direct3D 12 OpenXR backend and rendering with view
instancing (#86283). See godotengine/godot-proposals#10078.

This change adds support for running XR projects built with the `d3d12`
rendering backend. The XR backend hooks into the setup for the D3D12
render context in order to use the desired device and command queue for
submission to OpenXR. The XR backend takes care of importing the D3D12
swapchain images into the render context.

As part of this process, three issues are addressed:
- Ensuring that resource state transitions are only done on textures
  that require them.
- Enabling view instancing in the PSOs for multiview render passes.
- Addressing a bug in the D3D12 runtime where PSO creation may fail
  when front face detection is used.

Please refer to #86283 for additional discussions on the implementation
details.

Fixes https://github.com/godotengine/godot/issues/86283

----8<----8<----8<----8<----8<----8<----8<----8<----8<----8<----8<----
(remove content below from commit message when squashing this pull request)

For testing, I have been using the https://github.com/GodotVR/godot-xr-template and set it up for D3D12 rendering:

```
--- a/project.godot
+++ b/project.godot
@@ -54,8 +54,12 @@ common/drop_mouse_on_gui_input_disabled=true

 [rendering]

-renderer/rendering_method="gl_compatibility"
-renderer/rendering_method.mobile="gl_compatibility"
+rendering_device/driver="d3d12"
+rendering_device/driver.windows="d3d12"
+;rendering_device/d3d12/agility_sdk_version=615
+rendering_device/fallback_to_vulkan=false
+rendering_device/fallback_to_opengl3=false
+renderer/rendering_method.mobile="forward_plus"
 textures/vram_compression/import_etc2_astc=true
 environment/defaults/default_environment="res://default_env.tres"

```

Testing:
- Nvidia Ada Lovelace + godot-xr-template + Virtual Desktop (VDXR): PASS
- Nvidia Ada Lovelace + godot-xr-template + Varjo: PASS
- Nvidia Ada Lovelace + godot-xr-template + Meta XR Simulator: PASS
- Nvidia Ada Lovelace + godot-xr-template + Meta Quest Link: PASS
- Nvidia Ada Lovelace + godot-xr-template + SteamVR: PASS
- Nvidia Ada Lovelace + godot-xr-template + Windows Mixed Reality: FAIL [1]
- AMD RDNA2 + godot-xr-template + Virtual Desktop (VDXR): PASS
- Intel ARC + godot-xr-template + Virtual Desktop (VDXR): PASS
- godot-xr-template + Agility SDK 615: PASS
- godot-xr-template + `--gpu-validation`: PASS

[1] Disappointingly, there is an issue with WMR and we crash in `xrCreateSwapchain()` pretty inexplicably. This is not a bug with this change however.

mbucchia committed:
Add support for Direct3D 12 OpenXR backend.

This change adds support for running XR projects built with the `d3d12`
rendering backend. The XR backend hooks into the setup for the D3D12
render context in order to use the desired device and command queue for
submission to OpenXR. The XR backend takes care of importing the D3D12
swapchain images into the render context.

As part of this process, three issues are addressed:
- Ensuring that resource state transitions are only done on textures
  that require them.
- Enabling view instancing in the PSOs for multiview render passes.
- Addressing a bug in the D3D12 runtime where PSO creation may fail
  when front face detection is used.

Please refer to #86283 for additional discussions on the implementation
details.

mbucchia commented:
Looking into workflow failures that seem to be due to the use of `XrGraphicsRequirementsD3D12KHR`.

mbucchia commented:
I am a little confused by your CI.

This workflow with latest push passed all green:

https://github.com/mbucchia/godot/actions/runs/13881066387

However the checks above reflect and CSS error. The details show the formatter replacing `#ifdef` inclusion guards with `#pragma` once.

Meanwhile your CSS guide (https://docs.godotengine.org/en/latest/contributing/development/cpp_usage_guidelines.html#pragma-once-directive) says:

> #pragma once directive
> To follow the existing style, please use standard #ifdef-based include guards instead of #pragma once in new files

What am I missing?

mbucchia commented:
Oh I think I get it - the passing run is from my forked repo. And running the workflow on the Godot repo requires an approval, so the current results are not reflecting the latest push.

lawnjelly commented:
4.x very recently changed to `#pragma once` instead of header guards (#102298), the documentation may be out of date.

mbucchia commented:
> 4.x very recently changed to `#pragma once` instead of header guards (#102298), the documentation may be out of date.

Understood. I will change my new files to use the pragmas. Thanks!

BastiaanOlij commented:
Just giving this a first look and all seems pretty good. I don't know the DX12 backend well enough to give any insight on that so will need @clayjohn or @DarioSamo to verify the changes.

I'll find some time to do some testing on my end this week.

Re your comment here: https://github.com/godotengine/godot/issues/86283#issuecomment-2727254988
With WMR itself defuncts, do we want to put hacks like this in or just make note of them for anyone that wants to custom build for WMR?
I'm assuming here that SteamVR and other PC OpenXR runtimes will work with DX12 as well, the main goal here would be for those who wish to use the DX12 renderer on windows to be able to do VR with Godot.

mbucchia commented:
Thank you Bastiaan! Please suggest some more testing as well if you want. I saw a couple of XR demo samples, maybe I will give those a shot. If you have a stress test let me know.

> Re your comment here: [#86283 (comment)](https://github.com/godotengine/godot/issues/86283#issuecomment-2727254988)
> With WMR itself defuncts, do we want to put hacks like this in or just make note of them for anyone that wants to custom build for WMR?

I'm a little bit torn yeah. I think you really want to get rid of this warning message that calls specifically incompatibility with WMR today. D3D support was an avenue for that. 

But also as I mentioned, I don't know if the issue will remain specific to WMR _if_ a platform begins to use D3D12 as its internal gfx API in the future. There is none today AFAICT. I might talk some more to the D3D team at Microsoft about these scenarios of "mixed usage" of Agility SDK. Definitely nothing actionable on the current PR though.

BastiaanOlij reviewed (approved):
@mbucchia  Sorry it took me awhile to get to this and properly test it.

Functionally this is working great for me. I've tested it both with the demo project on my original issue and running OpenXR with DX12 over SteamVR.

Technically I'm going to trust that the DX12 implementation is sound. My only question is whether you added the `d3d12_hooks` logic because Vulkan had it, or if it actually is used? None of the methods are implemented so I think you can leave this out. 
In Vulkan this is required because we let OpenXR handle the Vulkan instance creation, but I don't think that is required for DX.

Regarding the issue around WMR mentioned in the original issue. I think we can ignore this. With WMR basically defunct and no longer available in Windows 11, adding any sort of workaround in the code is daft. If someone really still wants to work with WMR on Windows 10, they can follow the instructions on the issue and do a custom build of Godot.

@clayjohn might want to have a quick check on the DX code but I think this is pretty much good to merge depending on the answer to the hooks question.

mbucchia commented:
> Sorry it took me awhile to get to this and properly test it.
> 
> Functionally this is working great for me. I've tested it both with the demo project on my original issue and running OpenXR with DX12 over SteamVR.

No worries, and thanks for the testing!

> My only question is whether you added the `d3d12_hooks` logic because Vulkan had it, or if it actually is used? None of the methods are implemented so I think you can leave this out.
> In Vulkan this is required because we let OpenXR handle the Vulkan instance creation, but I don't think that is required for DX.

It's actually used and I did copy the general idea from Vulkan. But maybe there is a better way of doing it.

The methods are implemented in `openxr_d3d12_extension.cpp` since the class `OpenXRD3D12Extension` inherits from `D3D12Hooks`. I am using it to pass around the D3D12 command queue from the renderer initialization for example. See `rendering_device_driver_d3d12.cpp` L2351 for example. Is there a better way to store/capture the outcome of the renderer initialization otherwise?

BastiaanOlij commented:
> The methods are implemented in `openxr_d3d12_extension.cpp` since the class `OpenXRD3D12Extension` inherits from `D3D12Hooks`. I am using it to pass around the D3D12 command queue from the renderer initialization for example. See `rendering_device_driver_d3d12.cpp` L2351 for example. Is there a better way to store/capture the outcome of the renderer initialization otherwise?

Ah, that makes sense, totally missed that. Cool!

BastiaanOlij commented:
XR Meeting, we're happy for this to be merged as is, we can do any minor cleanup later.

mbucchia commented:
I meant to test on AMD and Intel GPU as well, give me maybe a day or two.

mbucchia commented:
I was able to test with an AMD RDNA2 card successfully. I am not able to make my Intel ARC card work today (nothing to do with the change or Godot, the card isn't enumerating in Windows). I will circle back later, not worth blocking merge IMO.

I rebased against latest upstream, rebuilt and ran a quick sanity test. Added stub method `get_density_map()` (like it is done in the OpenGL backend as well - though I will probably get back to proper support for that, since it looks interesting). Fixed the couple of blank spaces suggestions in the process.

I pushed my rebase in two times, 1st push is rebase only (no changes to the code, as there were no conflicts) and 2nd push is the small delta I've described above.

@BastiaanOlij, please go ahead and merge when ready. I recommend editing the final commit message to remove the bullet points of testing (see marker in my PR description, with the little scissors 8< 8<)

mbucchia commented on drivers/d3d12/rendering_device_driver_d3d12.cpp:
```diff
@@ -29,6 +29,7 @@
 
 #include "rendering_device_driver_d3d12.h"
+#include "d3d12_hooks.h"
 
```
Done.

mbucchia commented on drivers/d3d12/rendering_context_driver_d3d12.cpp:
```diff
@@ -29,6 +29,7 @@
 
 #include "rendering_context_driver_d3d12.h"
+#include "d3d12_hooks.h"
 
```
Done.

clayjohn commented:
We don't use the squash on merge tool in Github, so we usually ask contributors to squash their commits themselves before we merge. Can you go ahead and do that, then force push your squashed version?

mbucchia commented:
> We don't use the squash on merge tool in Github, so we usually ask contributors to squash their commits themselves before we merge. Can you go ahead and do that, then force push your squashed version?

Thank you, I've fixed up all commits and updated commit message to match the PR description (summary of the changes).

mbucchia commented:
Evidence that there is no delta between the two sets of commits:

```
PS D:\XR\godot> git diff  6f5522b..e3c215f
PS D:\XR\godot>
```

clayjohn commented:
Thanks!

BastiaanOlij commented:
> I am not able to make my Intel ARC card work today

I've had reports in the past that ARC is having difficulty with stereo rendering as is, if you have hardware that you can test at some point it be nice to know if things work with DirectX and ARC.

But don't see it as a reason not to merge, if there are issues we'd need to get Intel involved at some point, if they are willing.

mbucchia commented:
> I've had reports in the past that ARC is having difficulty with stereo rendering as is, if you have hardware that you can test at some point it be nice to know if things work with DirectX and ARC.

Alright, I assume it's a "picture or it didn't happen" moment :)

![godot](https://github.com/user-attachments/assets/612c0d40-60d9-47b6-a080-b197659b39f2)

```
[...]
2025-04-20 12:10:59 -0700: Device is: Oculus Quest2 (20)
2025-04-20 12:10:59 -0700: Recommended resolution: 2496x2592
2025-04-20 12:11:18 -0700: Using Direct3D 12 on adapter: Intel(R) Arc(TM) A310 LP Graphics
2025-04-20 12:11:18 -0700: Using D3D12 on adapter: Intel(R) Arc(TM) A310 LP Graphics
2025-04-20 12:11:28 -0700: Creating a swapchain with texture array
[...]
```

Working great with ARC, latest driver. No issue with the view instancing.

mbucchia commented:
Amending my previous comment, I am getting a D3D layer error when running with the ARC. It could be sign of a real issue:

```
D3D12 ERROR: ID3D12CommandList::OMSetRenderTargets: The RenderTargetView at slot 0 is not compatible with the DepthStencilView. When D3D12_FEATURE_DATA_D3D12_OPTIONS19::MismatchingOutputDimensionsSupported is FALSE, DepthStencilViews may only be used with RenderTargetViews if the effective dimensions of the Views are equal, as well as the Resource types, multisample count, and multisample quality. One exception is that DepthStencil width and/or height can be larger than RenderTarget width/height.The RenderTargetView at slot 0 has (w:2496,h:2592,as:2), while the Resource is a Texture2D with (mc:1,mq:0). The DepthStencilView has (w:2496,h:2592,as:1), while the Resource is a Texture2D with (mc:1,mq:0).  [ EXECUTION ERROR #728: SET_RENDER_TARGETS_INVALID]
```

The call in question is in `RenderingDeviceDriverD3D12::command_next_render_subpass()` (I confirmed it is not a runtime issue).

I'm thinking this is because the ARC driver doesn't support this `MismatchingOutputDimensionsSupported` while Nvidia does. That said, when you look at the details, things start to become odd:

The RenderTargetView at slot 0 has (w:2496,h:2592,as:2) [...]
The DepthStencilView has (w:2496,h:2592,as:1)

This looks like a real issue where maybe we are not allocating depth buffer for stereo. I still recommend to proceed with merging this change, and I have filed a follow-up issue here: #105599.

BastiaanOlij commented:
Thanks Matthieu, this is really great news, if DX works well on ARC including multiview, that may offer a good solution to those who were having problems with that in the past.

mbucchia commented:
@BastiaanOlij, what's next? I believe only approved contributors can merge (I don't see the option).

AFAICT we are good to do.

BastiaanOlij commented:
@mbucchia yeah, with the approvals it should just be a matter of the PR being merged when the next dev build is being prepared.

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
